msal/src/Microsoft.Identity.Client/PublicClientApplication.cs

```diff
@@ -95,7 +95,33 @@ public PublicClientApplication(string clientId, string authority)
             {
                 ClientId = clientId
             };
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="scopes"></param>
+        /// <param name="extraQueryParameters"></param>
+        /// <param name="deviceCodeResultCallback"></param>
+        /// <returns></returns>
+        public async Task<AuthenticationResult> AcquireTokenAsync(
+            IEnumerable<string> scopes, 
+            string extraQueryParameters, 
+            Action<DeviceCodeResult> deviceCodeResultCallback)

```

Comment:
>Action<DeviceCodeResult> deviceCodeResultCallback [](start = 12, length = 49)

not sure about this one, but it makes sense. right now though we do not allow developers control on the flow thus we should at least introduce a way to cancel the polling

Reply:
(btw. is an issue today in ADAL.NET)

---
In reply to: [219634068](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1263#discussion_r219634068) [](ancestors = 219634068)

Reply:
Async offers a rather clean way of providing cancellation, you just need to allow to pass in a cancellation token which the user can then call "Cancel()" on. In fact, it is recommened that all public async APIs offer an overload with a CancelationToken.